client/stripe-utils/utils.js

```diff
@@ -627,10 +615,17 @@ export const initializeUPEAppearance = ( api, isBlockCheckout = 'false' ) => {
 			? getStripeServerData()?.blocksAppearance
 			: getStripeServerData()?.appearance;
 
-	// If appearance is empty, get a fresh copy and save it in a transient.
+	const data = getStripeServerData();
+	const isValidContext =
+		isBlockCheckout === 'true' ||
+		( data.isCheckout && ! data.isOrderPay && ! data.isChangingPayment );
+
+	// If appearance is empty, only save it if in a valid context.
 	if ( ! appearance ) {
 		appearance = getAppearance( isBlockCheckout === 'true' );
-		api.saveAppearance( appearance, isBlockCheckout );
+		if ( isValidContext ) {
+			api.saveAppearance( appearance, isBlockCheckout );
+		}
```

Comment:
I think we should update the naming for `isValidContext` to `isValidUpdateContext`, as "isValidContext" implies that we're not in a valid context, which is pretty broad.
```suggestion
	const isValidUpdateContext =
		isBlockCheckout === 'true' ||
		( data.isCheckout && ! data.isOrderPay && ! data.isChangingPayment );

	if ( ! appearance ) {
		appearance = getAppearance( isBlockCheckout === 'true' );
		// If we have re-built the appearance, only update the settings in the checkout context
		if ( isValidUpdateContext ) {
			api.saveAppearance( appearance, isBlockCheckout );
		}
```

At a more subtle level, we only care about `isValidUpdateContext` if we have ref-fetched the content, so I have a slight preference for pushing the definition into that code block so the logic is only running when we need it. That's mostly from a code grokking perspective -- I don't like having to read through a complex assignment only to realise it is actually irrelevant later in the code. (This suggestion is non-blocking, to be clear.)

Reply:
Great suggestions. Github didn't allow me to apply your suggestion directly, so I did it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4297/commits/9c12069374d4911c8f9d3eaa1fd80280b9f297b3

Also moved the definition into the branch in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4297/commits/02805ab7b77020b54dae167f7b844f38a0e5b4b0